### PR TITLE
feat: call webhooks at least once

### LINKED ISSUE #5319: Async at least once delivery guarantee for quota notifications
As a ZITADEL administrator, I want quota notifications to be delivered at least once so no data is lost in error cases.

```[tasklist]
### Acceptance Criteria
- [ ] ZITADEL retries to send notifications as long as their webhook endpoint doesn't return a successful response code.
- [ ] ZITADEL sends notifications asynchronously
- [ ] ZITADEL exposes a metric endpoint with an in-memory counter about failed endpoint calls
- [ ] Implementation like guarant email notification
```

### PR DESCRIPTION
**Pull request by eliobischof**
Closes #5319 

# Status quo 
1. Quota threshold is reached
2. The corresponding webhook is called
3. If the call fails, it is retried twice
4. If successful, a *quota.NotifiedEvent* is emitted

# Changes in this PR
1. Quota threshold is reached
2. *quota.NotificationDueEvent* is emitted
3. The notification projection consumes *quota.NotificationDueEvents*
4. It does nothing further if it finds a *quota.NotifiedEvent* with the *quota.NotificationDueEvents* ID as a reference
5. It calls the webhook
6. It emits a *quota.NotifiedEvent* with the *quota.NotificationDueEvents* ID as a reference

# Unexpected benefits
- Notification payloads are also written to debug channels filesystem and stdout if enabled
- All notifications are instrumented with logging, metrics and tracing

# TODO
- [ ] ~unit tests~
- [x] fix e2e tests
- [x] add counter metric for failed calls

```[tasklist]
### Definition of Ready
- [x] I am happy with the code
- [x] Short description of the feature/issue is added in the pr description
- [x] PR is linked to the corresponding user story
- [x] Acceptance criteria are met
- [x] All open todos and follow ups are defined in a new ticket and justified
- [x] Deviations from the acceptance criteria and design are agreed with the PO and documented.
- [x] No debug or dead code
- [x] Critical parts are tested automatically
- [x] Where possible E2E tests are implemented
- [x] Documentation/examples are up-to-date
- [x] All non-functional requirements are met
- [ ] Functionality of the acceptance criteria is checked manually on the dev system.
```

### CODE REVIEW

File: internal/notification/senders/email.go [hunk 1]
```diff
@@ -5,15 +5,35 @@ import (
...
+	emailConfig func(ctx context.Context) (*smtp.Config, error),
+	getFileSystemProvider func(ctx context.Context) (*fs.Config, error),
+	getLogProvider func(ctx context.Context) (*log.Config, error),
+	successMetricName,
+	failureMetricName string,
+) (chain *Chain, err error) {
 	channels := make([]channels.NotificationChannel, 0, 3)
-	p, err := smtp.InitSMTPChannel(ctx, emailConfig)
+	p, err := smtp.InitChannel(ctx, emailConfig)
+	// TODO: Why is this error not handled?
```
Comment (eliobischof) [inline internal/notification/senders/email.go:25]: @hifabienne @livio-a any idea? It came with commit 7899a0b851a9fb3a056823756a9d6ac2eb88f690

File: internal/notification/senders/json.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+	getLogProvider func(ctx context.Context) (*log.Config, error),
+	successMetricName,
+	failureMetricName string,
+) (*Chain, error) {
+	if err := webhookConfig.Validate(); err != nil {
+		return nil, err
+	}
+	channels := make([]channels.NotificationChannel, 0, 3)
+	webhookChannel, err := webhook.InitChannel(ctx, webhookConfig)
+	// TODO: Handle this error?
```
Comment (eliobischof) [inline internal/notification/senders/json.go:28]: here too

File: internal/notification/channels/webhook/channel.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/zitadel/logging"
+
+	caos_errs "github.com/zitadel/zitadel/internal/errors"
```
Comment (adlerhurst) [inline internal/notification/channels/webhook/channel.go:12]: rename to errors

File: internal/notification/handlers/notifqry/config_filesystem.go [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+import (
+	"context"
+
+	"github.com/zitadel/zitadel/internal/api/authz"
+	"github.com/zitadel/zitadel/internal/domain"
+	"github.com/zitadel/zitadel/internal/notification/channels/fs"
+)
+
+// Read iam filesystem provider config
+func (n *NotificationQueries) GetFileSystemProvider(ctx context.Context) (*fs.Config, error) {
```
Comment (adlerhurst) [inline internal/notification/handlers/notifqry/config_filesystem.go:12]: use correct comment format

File: internal/notification/handlers/notifqry/config_log.go [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+import (
+	"context"
+
+	"github.com/zitadel/zitadel/internal/api/authz"
+	"github.com/zitadel/zitadel/internal/domain"
+	"github.com/zitadel/zitadel/internal/notification/channels/log"
+)
+
+// Read iam log provider config
+func (n *NotificationQueries) GetLogProvider(ctx context.Context) (*log.Config, error) {
```
Comment (adlerhurst) [inline internal/notification/handlers/notifqry/config_log.go:12]: use correct comment format

File: internal/notification/handlers/notifqry/config_twilio.go [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+	"github.com/zitadel/zitadel/internal/crypto"
+	"github.com/zitadel/zitadel/internal/domain"
+	"github.com/zitadel/zitadel/internal/errors"
+	"github.com/zitadel/zitadel/internal/notification/channels/twilio"
+	"github.com/zitadel/zitadel/internal/query"
+)
+
+// Read iam twilio config
+func (n *NotificationQueries) GetTwilioConfig(ctx context.Context) (*twilio.Config, error) {
```
Comment (adlerhurst) [inline internal/notification/handlers/notifqry/config_twilio.go:14]: use correct comment format

File: internal/notification/senders/email.go (hunk 1 above)
Comment (livio-a) [reply internal/notification/senders/email.go:25]: i mean it is handled (by ignoring it 😄 )... i think the idea was to be able to configure multiple channels where the message could be sent, but failing on one of them (including initialization) must not return an error

Comment (eliobischof) [reply internal/notification/senders/email.go:25]: Should we log it? At least a debug log?

File: internal/notification/handlers/notifqry/config_filesystem.go (hunk 4 above)
Comment (eliobischof) [reply internal/notification/handlers/notifqry/config_filesystem.go]: done

File: internal/notification/channels/webhook/channel.go (hunk 3 above)
Comment (eliobischof) [reply internal/notification/channels/webhook/channel.go]: done

**Review by adlerhurst — APPROVED** [review]

File: internal/notification/senders/email.go (hunk 1 above)
Comment (adlerhurst) [reply internal/notification/senders/email.go:25]: i would add a debug log

File: internal/notification/senders/json.go (hunk 2 above)
Comment (adlerhurst) [reply internal/notification/senders/json.go:28]: debug log :D

File: internal/notification/handlers/notifctx/notification.go [hunk 7]
```diff
@@ -0,0 +1,15 @@
+package notifctx
+
+import (
+	"context"
+
+	"github.com/zitadel/zitadel/internal/api/authz"
+	"github.com/zitadel/zitadel/internal/eventstore"
+)
+
+const NotifyUserID = "NOTIFICATION" //TODO: system?
```
[the remark this answers is not in the document]
Comment (adlerhurst) [reply internal/notification/handlers/notifctx/notification.go]: i would rework this as soon as we start redesign the aggregates

**Review by adlerhurst — APPROVED** [review]

File: internal/notification/senders/email.go (hunk 1 above)
Comment (eliobischof) [reply internal/notification/senders/email.go]: done

**Review by adlerhurst — APPROVED** [review]

**Pull request merged**